File: iOS/PauseletTests/AIImportControllerTests.swift
import XCTest
import ReminderCore
import ReminderAI
@testable import Pauselet

/// Tests for the iOS side of AI import: the key's life in the secret store,
/// and the guarantee that no request is made without one. The OpenAI client
/// and its parsing are covered by `ReminderAITests`; nothing here reaches the
/// network.
@MainActor
final class AIImportControllerTests: XCTestCase {

    private struct StubInterpreter: ExerciseInterpreting {
        let result: [Exercise]
        func interpret(_ text: String) async throws -> [Exercise] { result }
    }

    private struct FailingInterpreter: ExerciseInterpreting {
        let error: AIImportError
        func interpret(_ text: String) async throws -> [Exercise] { throw error }
    }

    /// Refuses every write, standing in for a keychain the app cannot reach.
    private struct UnwritableStore: SecretStoring {
        struct Denied: LocalizedError {
            var errorDescription: String? { "Could not reach the keychain." }
        }
        func read(account: String) throws -> String? { nil }
        func write(_ value: String?, account: String) throws { throw Denied() }
    }

    private func controller(
        secrets: SecretStoring = InMemorySecretStore(),
        interpreter: ExerciseInterpreting = StubInterpreter(result: [])
    ) -> AIImportController {
        AIImportController(secrets: secrets, makeInterpreter: { _, _ in interpreter })
    }

    // MARK: - The key

    func testStartsUnconfiguredWithNoStoredKey() {
        XCTAssertFalse(controller().isConfigured)
    }

    func testStoringAKeyConfiguresIt() {
        let ai = controller()

        XCTAssertNil(ai.store(key: "sk-test"))

        XCTAssertTrue(ai.isConfigured)
    }

    func testStoredKeyIsTrimmed() throws {
        let secrets = InMemorySecretStore()
        let ai = controller(secrets: secrets)

        ai.store(key: "  sk-test\n")

        XCTAssertEqual(try secrets.read(account: aiImportKeyAccount), "sk-test")
    }

    func testStoringNilRemovesTheKey() {
        let ai = controller(secrets: InMemorySecretStore([aiImportKeyAccount: "sk-test"]))
        XCTAssertTrue(ai.isConfigured)

        ai.store(key: nil)

        XCTAssertFalse(ai.isConfigured)
    }

    func testABlankKeyCountsAsNoKey() {
        let ai = controller()

        ai.store(key: "   ")

        XCTAssertFalse(ai.isConfigured)
    }

    /// A keychain that cannot be written to is worth saying out loud rather
    /// than silently losing what the user typed.
    func testAnUnwritableStoreReturnsAnError() {
        let ai = controller(secrets: UnwritableStore())

        let error = ai.store(key: "sk-test")

        XCTAssertNotNil(error)
        XCTAssertFalse(ai.isConfigured)
    }

    func testStoringAKeyClearsTheLastTestResult() {
        let ai = controller()
        ai.testResult = .failure("stale")

        ai.store(key: "sk-test")

        XCTAssertNil(ai.testResult)
    }

    // MARK: - Interpreting

    func testInterpretingWithoutAKeyThrowsRatherThanCallingOut() async {
        let ai = controller(
            interpreter: StubInterpreter(result: [Exercise(name: "Should not happen")])
        )

        do {
            _ = try await ai.interpret("3 sets of 10 squats")
            XCTFail("expected missingKey")
        } catch {
            XCTAssertEqual(error as? AIImportError, .missingKey)
        }
    }

    func testInterpretingWithAKeyReturnsTheInterpretedExercises() async throws {
        let ai = controller(
            secrets: InMemorySecretStore([aiImportKeyAccount: "sk-test"]),
            interpreter: StubInterpreter(result: [Exercise(name: "Squats", sets: 3, reps: 10)])
        )

        let exercises = try await ai.interpret("3 sets of 10 squats")

        XCTAssertEqual(exercises.map(\.name), ["Squats"])
    }

    // MARK: - Testing the key

    func testTestKeySucceedsWhenTheInterpreterAnswers() async {
        let ai = controller(
            secrets: InMemorySecretStore([aiImportKeyAccount: "sk-test"]),
            interpreter: StubInterpreter(result: [Exercise(name: "Squats")])
        )

        ai.testKey()
        await settle(ai)

        XCTAssertEqual(ai.testResult, .success)
    }

    /// The key and the model worked; the model just found nothing in a
    /// deliberately trivial phrase. That is still a pass.
    func testNothingFoundStillCountsAsAWorkingKey() async {
        let ai = controller(
            secrets: InMemorySecretStore([aiImportKeyAccount: "sk-test"]),
            interpreter: FailingInterpreter(error: .nothingFound)
        )

        ai.testKey()
        await settle(ai)

        XCTAssertEqual(ai.testResult, .success)
    }

    func testABadKeyReportsTheFailure() async {
        let ai = controller(
            secrets: InMemorySecretStore([aiImportKeyAccount: "sk-bad"]),
            interpreter: FailingInterpreter(error: .unauthorized)
        )

        ai.testKey()
        await settle(ai)

        guard case .failure = ai.testResult else {
            return XCTFail("expected a failure, got \(String(describing: ai.testResult))")
        }
    }

    /// `testKey` runs its work in a Task; wait for it to report finished
    /// rather than sleeping for a fixed time. `isTesting` is the completion
    /// signal — `testResult` is nil both before the work starts and, briefly,
    /// while it runs, so polling that instead races the failure path.
    private func settle(
        _ ai: AIImportController, file: StaticString = #filePath, line: UInt = #line
    ) async {
        for _ in 0..<1000 where ai.isTesting {
            await Task.yield()
        }
        XCTAssertFalse(ai.isTesting, "testKey never finished", file: file, line: line)
    }
}

File: Windows/Pauselet.Core.Tests/AIImportControllerTests.cs
using Pauselet.Core;
using Xunit;

namespace Pauselet.Core.Tests;

/// <summary>
/// The API key's life in the secret store, and the guarantee that nothing
/// reaches OpenAI without one. (Mirrors AIImportControllerTests on iOS.)
/// </summary>
public class AIImportControllerTests
{
    /// <summary>
    /// Records what it was asked to interpret, so a test can prove a request
    /// was — or was not — made.
    /// </summary>
    private sealed class RecordingInterpreter : IExerciseInterpreter
    {
        public List<string> Requests { get; } = [];
        public string? KeyUsed { get; set; }
        public AIImportModel? ModelUsed { get; set; }
        public Exception? Failure { get; set; }

        public Task<IReadOnlyList<Exercise>> InterpretAsync(
            string text, CancellationToken cancellationToken = default)
        {
            Requests.Add(text);
            if (Failure is not null) throw Failure;
            return Task.FromResult(ExerciseImporter.Parse(text));
        }
    }

    private static (AIImportController Controller, RecordingInterpreter Interpreter,
        InMemorySecretStore Secrets) Build(string? initialKey = null)
    {
        var secrets = new InMemorySecretStore();
        if (initialKey is not null) secrets.Write(initialKey, SecretAccounts.AIImportKey);
        var interpreter = new RecordingInterpreter();
        var controller = new AIImportController(secrets, (key, model) =>
        {
            interpreter.KeyUsed = key;
            interpreter.ModelUsed = model;
            return interpreter;
        });
        return (controller, interpreter, secrets);
    }

    // MARK: - The key

    [Fact]
    public void StartsUnconfiguredWithNoKey()
    {
        var (controller, _, _) = Build();
        Assert.False(controller.IsConfigured);
    }

    [Fact]
    public void StartsConfiguredWhenAKeyIsAlreadyStored()
    {
        var (controller, _, _) = Build("sk-existing");
        Assert.True(controller.IsConfigured);
    }

    [Fact]
    public void StoringAKeyConfiguresTheController()
    {
        var (controller, _, secrets) = Build();
        Assert.Null(controller.Store("sk-test"));
        Assert.True(controller.IsConfigured);
        Assert.Equal("sk-test", secrets.Read(SecretAccounts.AIImportKey));
    }

    [Fact]
    public void StoringABlankKeyRemovesIt()
    {
        var (controller, _, secrets) = Build("sk-existing");
        controller.Store("   ");
        Assert.False(controller.IsConfigured);
        Assert.Null(secrets.Read(SecretAccounts.AIImportKey));
    }

    [Fact]
    public void StoringNullRemovesTheKey()
    {
        var (controller, _, secrets) = Build("sk-existing");
        controller.Store(null);
        Assert.False(controller.IsConfigured);
        Assert.Null(secrets.Read(SecretAccounts.AIImportKey));
    }

    [Fact]
    public void SurroundingWhitespaceIsTrimmedOffAPastedKey()
    {
        var (controller, _, secrets) = Build();
        controller.Store("  sk-pasted\n");
        Assert.Equal("sk-pasted", secrets.Read(SecretAccounts.AIImportKey));
    }

    // MARK: - Using it

    [Fact]
    public async Task InterpretingWithoutAKeyMakesNoRequest()
    {
        var (controller, interpreter, _) = Build();
        var exception = await Assert.ThrowsAsync<AIImportException>(
            () => controller.InterpretAsync("3 sets of 10 squats"));
        Assert.Equal(AIImportException.Reason.MissingKey, exception.FailureReason);
        Assert.Empty(interpreter.Requests);
    }

    [Fact]
    public async Task InterpretingPassesTheStoredKeyAndChosenModel()
    {
        var (controller, interpreter, _) = Build("sk-test");
        controller.Model = AIImportModel.Mini;
        var exercises = await controller.InterpretAsync("Chin tucks 3 x 10, hold 5 seconds");

        Assert.Equal("sk-test", interpreter.KeyUsed);
        Assert.Equal(AIImportModel.Mini, interpreter.ModelUsed);
        Assert.Single(interpreter.Requests);
        Assert.Equal("Chin tucks", exercises[0].Name);
    }

    [Fact]
    public async Task ATestKeyRunReportsSuccess()
    {
        var (controller, _, _) = Build("sk-test");
        await controller.TestKeyAsync();
        Assert.True(controller.LastTestResult?.Succeeded);
        Assert.False(controller.IsTesting);
    }

    /// <summary>
    /// The key and the model worked; the model just found nothing in a
    /// deliberately trivial phrase. That is still a pass.
    /// </summary>
    [Fact]
    public async Task NothingFoundStillCountsAsAWorkingKey()
    {
        var (controller, interpreter, _) = Build("sk-test");
        interpreter.Failure = AIImportException.NothingFound();
        await controller.TestKeyAsync();
        Assert.True(controller.LastTestResult?.Succeeded);
    }

    [Fact]
    public async Task ARejectedKeyIsReportedWithItsMessage()
    {
        var (controller, interpreter, _) = Build("sk-bad");
        interpreter.Failure = AIImportException.Unauthorized();
        await controller.TestKeyAsync();
        Assert.False(controller.LastTestResult?.Succeeded);
        Assert.Equal(
            "That API key was rejected. Check it in Settings.",
            controller.LastTestResult?.Message);
    }

    [Fact]
    public void StoringANewKeyClearsTheLastTestResult()
    {
        var (controller, _, _) = Build("sk-test");
        controller.Store("sk-other");
        Assert.Null(controller.LastTestResult);
    }
}
